### PR TITLE
[FLINK-16094][docs] Translate /dev/table/functions/udfs.zh.md

### PR DESCRIPTION
**Pull request by libenchao**
## What is the purpose of the change

Translate /dev/table/functions/udfs.zh.md

## Brief change log

Translate /dev/table/functions/udfs.zh.md

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by xintongsong** [comment]
Hi @libenchao,

It seems this PR has diverged quite a lot from the latest master branch. This is obviously due to this PR has not been reviewed timely, and I'm sorry for your bad contribution experience.

Would you like to rebase and update the translations? If you would, I can help with the review this time.

**Comment by libenchao** [comment]
@xintongsong Thanks for your attention, I'll do a rebase and fix the conflicts soon.

**Comment by libenchao** [comment]
@xintongsong I've rebased to the master branch.

**Comment by xintongsong** [comment]
Thanks @libenchao, I'm reviewing this right now.

File: docs/dev/table/functions/udfs.zh.md [hunk 1]
```diff
@@ -603,15 +599,15 @@ abstract class AggregateFunction[T, ACC] extends UserDefinedAggregateFunction[T,
...
+为了计算加权平均值，accumulator 需要存储加权总和以及数据的条数。在我们的例子里，我们定义了一个类 `WeightedAvgAccum` 来作为 accumulator。Flink 的 checkpoint 机制会自动保存 accumulator，在失败时进行恢复，以此来保证精确一次的语义。
 
-The `accumulate()` method of our `WeightedAvg` `AggregateFunction` has three inputs. The first one is the `WeightedAvgAccum` accumulator, the other two are user-defined inputs: input value `ivalue` and weight of the input `iweight`. Although the `retract()`, `merge()`, and `resetAccumulator()` methods are not mandatory for most aggregation types, we provide them below as examples. Please note that we used Java primitive types and defined `getResultType()` and `getAccumulatorType()` methods in the Scala example because Flink type extraction does not work very well for Scala types.
+我们的 `WeightedAvg` 的 `accumulate` 方法有三个输入参数。第一个是 `WeightedAvgAccum` accumulator，另外两个是用户自定义的输入：输入的值 `ivalue` 和 输入的权重 `iweight`。尽管 `retract()`、`merge()`、`resetAccumulator()` 这几个方法在大多数聚合类型中都不是必须实现的，我们也在样例中提供了他们的实现。请注意我们在 Scala 样例中也是用的是 Java 的基础类型，并且定义了 `getResultType()` 和 `getAccumulatorType()`，因为 Flink 的类型推导对于 Scala 的类型推导做的不是很好。
```
Comment (xintongsong) [inline docs/dev/table/functions/udfs.zh.md:610]: “我们的 `WeightedAvg`” 初次提到 `WeightedAvg` 应该说明它是一个 `AggregateFunction`

File: docs/dev/table/functions/udfs.zh.md [hunk 2]
```diff
@@ -826,46 +823,44 @@ t_env.sql_query("SELECT user, wAvg(points, level) AS avgPoints FROM userScores G
...
 
-User-Defined Table Aggregate Functions (UDTAGGs) aggregate a table (one or more rows with one or more attributes) to a result table with multi rows and columns. 
+自定义表值聚合函数（UDTAGG）可以把一个表（一行或者多行，每行有一列或者多列）聚合成另一张表，结果中可以有多行多列。
 
 <center>
 <img alt="UDAGG mechanism" src="{{ site.baseurl }}/fig/udtagg-mechanism.png" width="80%">
 </center>
 
-The above figure shows an example of a table aggregation. Assume you have a table that contains data about beverages. The table consists of three columns, `id`, `name` and `price` and 5 rows. Imagine you need to find the top 2 highest prices of all beverages in the table, i.e., perform a `top2()` table aggregation. You would need to check each of the 5 rows and the result would be a table with the top 2 values.
+上图展示了一个表值聚合函数的例子。假设你有一个饮料的表，这个表有3列，分别是 `id`、`name` 和 `price`，一共有5行。假设你需要找到价格最高的两个饮料，类似于 `top2()` 表值聚合函数。你需要遍历所有5行数据，结果是有2行数据的一个表。
```
Comment (xintongsong) [inline docs/dev/table/functions/udfs.zh.md:835]:
```suggestion
上图展示了一个表值聚合函数的例子。假设你有一个饮料的表，这个表有 3 列，分别是 `id`、`name` 和 `price`，一共有 5 行。假设你需要找到价格最高的两个饮料，类似于 `top2()` 表值聚合函数。你需要遍历所有 5 行数据，结果是有 2 行数据的一个表。
```

File: docs/dev/table/functions/udfs.zh.md [hunk 3]
```diff
@@ -1126,15 +1121,15 @@ abstract class TableAggregateFunction[T, ACC] extends UserDefinedAggregateFuncti
 </div>
 
 
-The following example shows how to
+下面的例子展示了如何
 
-- define a `TableAggregateFunction` that calculates the top 2 values on a given column, 
-- register the function in the `TableEnvironment`, and 
-- use the function in a Table API query(TableAggregateFunction is only supported by Table API).  
+- 定义一个 `TableAggregateFunction` 来计算最大的2个值，
```
Comment (xintongsong) [inline docs/dev/table/functions/udfs.zh.md:1126]: "of a given column"

Comment (xintongsong) [reply docs/dev/table/functions/udfs.zh.md:1126]:
```suggestion
- 定义一个 `TableAggregateFunction` 来计算最大的 2 个值，
```

File: docs/dev/table/functions/udfs.zh.md [hunk 4]
```diff
@@ -1425,33 +1420,33 @@ tab
...
 ---------------------------------
 
-Sometimes it might be necessary for a user-defined function to get global runtime information or do some setup/clean-up work before the actual work. User-defined functions provide `open()` and `close()` methods that can be overridden and provide similar functionality as the methods in `RichFunction` of DataSet or DataStream API.
+有时候自定义函数需要获取一些全局信息，或者在真正被调用之前做一些配置（setup）/清理（clean-up）的工作。自定义函数也提供了 `open()` 和 `close()` 方法，你可以重写这两个方法做到类似于 DataSet 或者 DataStream API 中 `RichFunction` 的功能。
 
-The `open()` method is called once before the evaluation method. The `close()` method after the last call to the evaluation method.
+`open()` 方法在自定义函数被调用之前先调用。`close()` 方法在自定义函数调用完之后被调用。
 
-The `open()` method provides a `FunctionContext` that contains information about the context in which user-defined functions are executed, such as the metric group, the distributed cache files, or the global job parameters.
+`open()` 方法提供了一个 `FunctionContext`，它包含了一些自定义函数被执行时的上下文信息，比如 metric group、分布式文件缓存，或者是全局的任务参数等。
```
Comment (xintongsong) [inline docs/dev/table/functions/udfs.zh.md:1439]: “任务” -> “作业”

File: docs/dev/table/functions/udfs.zh.md [hunk 5]
```diff
@@ -1425,33 +1420,33 @@ tab
...
+下面的信息可以通过调用 `FunctionContext` 的对应的方法来获得：
 
-| Method                                | Description                                            |
+| 方法                                  | 描述                                                    |
 | :------------------------------------ | :----------------------------------------------------- |
-| `getMetricGroup()`                    | Metric group for this parallel subtask.                |
-| `getCachedFile(name)`                 | Local temporary file copy of a distributed cache file. |
-| `getJobParameter(name, defaultValue)` | Global job parameter value associated with given key.  |
+| `getMetricGroup()`                    | 执行该函数的 subtask 的 Metric Group。                   |
+| `getCachedFile(name)`                 | 分布式文件缓存的本地临时文件拷贝。                         |
```
Comment (xintongsong) [inline docs/dev/table/functions/udfs.zh.md:1446]:
```suggestion
| `getCachedFile(name)`                 | 分布式文件缓存的本地临时文件副本。                         |
```

File: docs/dev/table/functions/udfs.zh.md [hunk 6]
```diff
@@ -271,32 +269,32 @@ env = StreamExecutionEnvironment.get_execution_environment()
...
 
-There are many ways to define a Python table function besides extending the base class `TableFunction`.
-Please refer to the [Python Table Function]({{ site.baseurl }}/zh/dev/table/python/python_udfs.html#table-functions) documentation for more details.
+除了继承 `TableFunction`，还有很多其它方法可以定义 Python 表值函数。
+更多信息，参考 [Python 表值函数]({{ site.baseurl }}/zh/dev/table/python/python_udfs.html#table-functions)文档。
 
 </div>
 </div>
 
 <div class="codetabs" markdown="1">
```
[the remark this answers is not in the document]
Comment (wuchong) [reply docs/dev/table/functions/udfs.zh.md:291]: Sounds good to me.

File: docs/dev/table/functions/udfs.zh.md [hunk 7]
```diff
@@ -346,40 +344,38 @@ class CustomTypeSplit extends TableFunction[Row] {
...
- `AggregateFunction#getResultType()`) and the type of the accumulator (through `AggregateFunction#getAccumulatorType()`).
+Flink 的类型推导不能处理复杂的数据类型，只能处理基础类型或者是简单的 POJO 类型。所以跟 `ScalarFunction` 和 `TableFunction` 一样，`AggregateFunction` 也提供了 `AggregateFunction#getResultType()` 和 `AggregateFunction#getAccumulatorType()` 来分别指定返回值类型和 accumulator 的类型，两个函数的返回值类型也都是 `TypeInformation`。
  
-Besides the above methods, there are a few contracted methods that can be 
-optionally implemented. While some of these methods allow the system more efficient query execution, others are mandatory for certain use cases. For instance, the `merge()` method is mandatory if the aggregation function should be applied in the context of a session group window (the accumulators of two session windows need to be joined when a row is observed that "connects" them). 
+除了上面的方法，还有几个方法可以选择实现。这些方法有些可以让查询更加高效，而有些是在某些特定场景下必须要实现的。例如，如果聚合函数用在会话窗口（当两个会话窗口合并的时候需要 merge 他们的 accumulator）的话，`merge()` 方法就是必须要实现的。
```
[the remark this answers is not in the document]
Comment (wuchong) [reply docs/dev/table/functions/udfs.zh.md:368]: 我觉得这里是意译了，本意和原文是一致的，如本超所说，原文意思是“两个本来没有交集的window，可能因为一个元素的到来被merge起来”，其实和翻译“当两个会话窗口合并的时候”要表达的内容是差不多的，本超这里的翻译更简洁些。所以我觉得可以的。

[the remark this answers is not in the document]
Comment (xintongsong) [reply docs/dev/table/functions/udfs.zh.md:368]: Sounds good. Thanks for the explanations.

**Comment by libenchao** [comment]
@xintongsong Thanks a lot for the review. I've addressed all your comments and do a refactor with a hotfix commit.
@wuchong I'm not very familiar with the merging process, if you need I do the rebase before merging, pls let me know. (Or if you think this should be another issue, I'll open another pr to do that)

**Comment by libenchao** [comment]
@wuchong Thanks for the reminder and merging.

**Pull request closed without merging**